File: public/code/smc__liquidity_grab_bos.ts
import { Candle, Signal, makeSignal, ema, atr, swingHighs, swingLows } from "../indicators";

export function liquidityGrabBos(c: Candle[]): Signal {
  if (c.length < 30) return makeSignal({ reason: "Insufficient data" });
  const sh = swingHighs(c, 10).filter((v): v is number => v !== null);
  const sl = swingLows(c, 10).filter((v): v is number => v !== null);
  if (sh.length < 2 || sl.length < 2) return makeSignal({ reason: "No clean structure" });
  const a = atr(c, 14);
  const cur = c[c.length - 1], prev = c[c.length - 2];
  const lastH = sh[sh.length - 1], lastL = sl[sl.length - 1];
  if (prev.low < lastL && prev.close > lastL && cur.close > lastH) {
    const slv = prev.low - 0.3 * a[a.length - 1], r = cur.close - slv;
    return makeSignal({ signal: "long", entry: cur.close, stop_loss: slv, take_profit: [cur.close + r * 2, cur.close + r * 3, cur.close + r * 5], confidence: 0.82, reason: "Liquidity grab + BOS up" });
  }
  if (prev.high > lastH && prev.close < lastH && cur.close < lastL) {
    const slv = prev.high + 0.3 * a[a.length - 1], r = slv - cur.close;
    return makeSignal({ signal: "short", entry: cur.close, stop_loss: slv, take_profit: [cur.close - r * 2, cur.close - r * 3, cur.close - r * 5], confidence: 0.82, reason: "Liquidity grab + BOS down" });
  }
  return makeSignal({ reason: "LG+BOS kurulumu yok" });
}

File: public/code/smc__liquidity_grab_bos.py
from common import make_signal, ema, sma, rsi, macd, atr, bollinger_bands, vwap, swing_highs, swing_lows
import math


def _sign(x):
    return (x > 0) - (x < 0)


def liquidityGrabBos(c):
    if len(c) < 30:
        return make_signal(reason="Insufficient data")
    sh = [v for v in swingHighs(c, 10) if v != None]
    sl = [v for v in swingLows(c, 10) if v != None]
    if len(sh) < 2  or  len(sl) < 2:
        return make_signal(reason="No clean structure")
    a = atr(c, 14)
    cur = c[len(c) - 1]
    prev = c[len(c) - 2]
    lastH = sh[len(sh) - 1]
    lastL = sl[len(sl) - 1]
    if prev['low'] < lastL  and  prev['close'] > lastL  and  cur['close'] > lastH:
        slv = prev['low'] - 0.3 * a[len(a) - 1]
        r = cur['close'] - slv
        return make_signal(signal="long", entry=cur['close'], stop_loss=slv, take_profit=[cur['close'] + r * 2, cur['close'] + r * 3, cur['close'] + r * 5], confidence=0.82, reason="Liquidity grab + BOS up")
    if prev['high'] > lastH  and  prev['close'] < lastH  and  cur['close'] < lastL:
        slv = prev['high'] + 0.3 * a[len(a) - 1]
        r = slv - cur['close']
        return make_signal(signal="short", entry=cur['close'], stop_loss=slv, take_profit=[cur['close'] - r * 2, cur['close'] - r * 3, cur['close'] - r * 5], confidence=0.82, reason="Liquidity grab + BOS down")
    return make_signal(reason="LG+BOS kurulumu yok")
